File: src/components/installers/neovim.ts
import * as path from 'path';
import { TerraformStack } from 'cdktf';
import { Construct } from 'constructs';
import { Aptitude } from './aptitude';
import { Brew } from './brew';
import { GitHubRepoInstaller } from './github';
import { Platform, Project } from '../../core';


export class Neovim extends Construct {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    const commands = [
      'rm -rf $HOME/neovim/share/nvim/runtime',
      'make CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/neovim" CMAKE_BUILD_TYPE=Release',
      'make install',
    ];

    const repo = new GitHubRepoInstaller(this, 'neovim', {
      org: 'neovim',
      repo: 'neovim',
      folderName: 'neovim-install',
      installCommands: commands,
      updateCommands: commands,
    });

    const neovimDeps = new TerraformStack(this, 'neovim-deps');
    repo.addDependency(neovimDeps);

    if (Project.ofProject(this).platform === Platform.LINUX) {
      [
        'ninja-build',
        'gettext libtool',
        'libtool-bin',
        'autoconf',
        'automake',
        'cmake',
        'g++',
        'pkg-config',
        'unzip',
        'curl',
        'doxygen',
      ].forEach(dep => {
        new Aptitude(neovimDeps, dep, {
          pkgName: dep,
        });
      });
    } else if (Project.ofProject(this).platform === Platform.DARWIN) {
      [
        'ninja',
        'libtool',
        'automake',
        'cmake',
        'pkg-config',
        'gettext',
        'curl',
      ].forEach(dep => {
        new Brew(neovimDeps, dep, {
          pkgName: dep,
        });
      });
    }

    const project = Project.ofProject(this);
    project.profile.addToSystemPath(path.join(project.home.homeVar, 'neovim', 'bin'));
  }
}

File: pkg/components/neovim.go
package components

import (
	"fmt"
	"path"

	"github.com/corymhall/pulumi-provider-pde/sdk/go/pde/installers"
	"github.com/pulumi/pulumi-github/sdk/v6/go/github"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

type Neovim interface {
	Component
}

type neovim struct {
	pulumi.ResourceState
	*component
}

func NewNeovim(ctx *pulumi.Context, project *Project, name string, opts pulumi.ResourceOption) (Neovim, error) {
	n := &neovim{
		component: NewComponent(),
	}
	if err := ctx.RegisterComponentResource("pde:index:Neovim", name, n, opts); err != nil {
		return nil, err
	}

	ref, err := github.GetRef(ctx, &github.GetRefArgs{
		Owner:      pulumi.StringRef("neovim"),
		Repository: "neovim",
		Ref:        "tags/v0.10.1",
	})
	if err != nil {
		return nil, err
	}

	// releaseCommands := pulumi.StringArray{
	// 	pulumi.String("xattr -c ./nvim-macos.tar.gz"),
	// 	pulumi.String("tar xzvf nvim-macos.tar.gz"),
	// 	pulumi.String("rm -rf $HOME/neovim"),
	// 	pulumi.String("mv nvim-macos $HOME/neovim"),
	// }
	installCommands := pulumi.ToStringArray([]string{
		fmt.Sprintf("rm -rf %s/neovim/share/nvim/runtime || true", project.Home.HomeLocation),
		`rm -r build/ || true`,
		`make CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=$HOME/neovim" CMAKE_BUILD_TYPE=Release`,
		"make install",
	})
	_, err = installers.NewGitHubRepo(ctx, "neovim", &installers.GitHubRepoArgs{
		Org:             pulumi.String("neovim"),
		Repo:            pulumi.String("neovim"),
		FolderName:      pulumi.String("neovim-install"),
		Version:         pulumi.String(ref.Sha),
		InstallCommands: installCommands,
		UpdateCommands:  installCommands,
		UninstallCommands: pulumi.ToStringArray([]string{
			fmt.Sprintf("rm -rf %s/neovim", project.Home.HomeLocation),
			fmt.Sprintf("rm -rf %s/neovim-install", project.Home.HomeLocation),
		}),
	})
	if err != nil {
		return nil, err
	}

	n.AddDeps(
		BrewPackage("ninja"),
		BrewPackage("cmake"),
		BrewPackage("gettext"),
		BrewPackage("curl"),
	)

	n.AddToSystemPath(path.Join(project.Home.HomeVar, "neovim", "bin"))

	return n, nil
}
